src/components/Search/CompareWithBase.tsx

```diff
@@ -70,23 +70,19 @@ function CompareWithBase({
   const [expanded, setExpanded] = useState(true);
   const { enqueueSnackbar } = useSnackbar();
 
-  //the "committed" base and new revisions initialize the staging state
+  // The "committed" base and new revisions initialize the staging state.
+  // These states are snapshots of selections, that the user can either save (by
+  // pressing the "Save" button) or revert to (by pressing the Cancel button).
   const [baseStagingRevs, setStagingBaseRevs] = useState<Changeset[]>(baseRevs);
-
   const [newStagingRevs, setStagingNewRevs] = useState<Changeset[]>(newRevs);
 
-  //the edit button will initialize the "in progress" state
-  //and copy "stage" to "in progress" state
-  const [baseInProgressRevs, setInProgressBaseRevs] = useState<Changeset[]>([]);
-  const [baseInProgress, setInProgressBase] = useState(false);
-
-  const [newInProgressRevs, setInProgressNewRevs] = useState<Changeset[]>([]);
-  const [newInProgress, setInProgressNew] = useState(false);
-
-  const [displayedRevisionsBaseRevs, setDisplayedRevisionsBaseRevs] =
+  // The "committed" base and new revisions initialize the "in progress" state
+  // too. These states hold the data that are displayed to the user. They always
+  // contain the displayed data. That's also the ones that will be committed if
+  // the user presses the "Compare" button.
```

Comment:
I thought the "staging" states become "committed" when the user presses the "Compare" button?

Shouldn't pressing "Save" be a requirement before pressing the "Compare" button ? Unless you changed that logic too? I remembered we talked about how unnecessary that extra step was. Did you change it?

When the user presses the "Save" button, the "inprogress" states overwrite the "staging" states? 

Also, this bring up another question. Should the "Compare" button be active once the results page mounts? Wouldn't it make sense for it to be active when the user presses the "Edit" button?

Reply:
Thanks for the great questions! I can see where the confusion comes from.

My goal for this patch is that no behavior was to be changed. We discussed in the past that the "Staging" state was  unnecessary and I'm all for it, but wanted to defer any behavior change to after this refactoring.


> I thought the "staging" states become "committed" when the user presses the "Compare" button?
> 
> Shouldn't pressing "Save" be a requirement before pressing the "Compare" button ? Unless you changed that logic too? I remembered we talked about how unnecessary that extra step was. Did you change it?

I didn't change this in _this_ patch, but indeed I changed it in a previous patch. I think that was in #582, more specifically in the change to `SelectedRevisionItem.tsx` in https://github.com/mozilla/perfcompare/pull/582/commits/cd04c4aed944171efb71cd75908b017d8f3646c6#diff-3db985901d6838f35e9001408344d9ed36d6c3377ee8e0ccb8d0d0f38bad9c29R64: the revisions that are displayed are the ones that are submitted when the user presses `Compare`.

Now that I think about it, I should probably have added these hidden fields in `CompareWithBase` so that it would be easier to control what is actually submitted; and therefore I could have used the "staging" data instead of the "in progress" data.

So currently, the "staging" data is only used as a snapshot (at least that's how I think about it): the user can "edit", then press "cancel" to go back at the previously "save"d snapshot. Or "save" to save a snapshot. I believe it's not so useful, especially that with the router now, the user can press the browser's back button after "compare" to come back at the previous state.

I'd like to not change this behavior in this patch, but for the near future there are IMO 2 possible paths (both of them in another patch though):

1. I'd be fine reverting this to the previous approach (i.e. "Staging" is used when the user presses "Compare"), to stay faithful to my previous goal "do no change behavior just yet", so that we can discuss this further.
2. Remove both the staging state and the "save" button (keeping Edit button though).

My preference is solution 2.

> 
> When the user presses the "Save" button, the "inprogress" states overwrite the "staging" states?

Yes!

> 
> Also, this bring up another question. Should the "Compare" button be active once the results page mounts? Wouldn't it make sense for it to be active when the user presses the "Edit" button?

Yeah, definitely, it makes sense that it is active only after the user did at least one edit.

I can file a JIRA ticket for this and the previous suggestion, actually I believe it makes sense that they are both grouped in one patch.

Tell me what you prefer!

Reply:
This sounds good for now. Yes, I agree. Let's defer any new changes to a new patch so it's clearer. Thank you!

Reply:
I'm also drawn to solution 2.